study_guide/questions.md

```diff
@@ -0,0 +1,179 @@
+- What's direct initialization of an object?
+- What's assignment initialization of an object?
+- Explain each constructor/destructor call caused by the sentence in line 35 here: https://godbolt.org/z/5hPd6o .
+- Consider this piece of code: https://godbolt.org/z/PeW6Tv . What's the order in which objects `base_va_`, `base_vb_`, `derived_va_`, `derived_vb`, and the constructors of `Base` and `Derived` get executed? What the destruction order for the same items?
+- What construction/destruction order do member variables have within a class?
+- What's the constructor initialization list is? how does it affect member initialization order?
+- What's the difference between initializing something in the initialization list of the constructor and initializing it within the body of the constructor?
+- How does the "-fno-elide-constructors" argument affect the outcome in https://godbolt.org/z/5hPd6o ?
+- What's a delegating constructor?
+
+== Instance methods and data members.
+
+- What's a static method? How does it differ from an instance method?
+- How do you call a static method? and an instance method?
+- What's a static variable? how does it differ from an instance member?
+
+== Overloading
+
+- What's overloading?
+- What's the function/method signature? What's in and what's out?
+- How does the compiler choose what overload to use for a particular call of an overloaded funciton?
+- What's the difference between "int value();" and "int value() const;"? Can you have both in a class?
+- What's operator overloading?
+- What operators can be overloaded?
+- When should you use operator overloading?
+- what's the signature to overload the operator "+"?
+- what's the signature to overload the operator "=+"?
+
+== Move semantics
+
+- What's an rvalue? provide examples.
+- What's an lvalue? provide examples.
+- How does `std::move()` move an object into another, if it can't access the target?  
+- What is a copy constructor? What's the signature?
+- What is a move constructor? What's the signature?
+- What is a copy assingment? What's the signature?
+- What is a move assignment? What's the signature?
+- Consider "const std::string value();" How does `const` prevent using move semantics?
+- What's the rule of three?
+- What's the rule of five?
+- Under what circumstances are default copy operations created by the compiler?
+- Under what circumstances are default move operations created by the compiler?
+
+== Containers
+
+- What's an `std::array`?
+- What's an `std::vector`?
+- What's an `std::list`?
+- What's an `std::map`? What's an `std::unordered_map`?
+- What's an `std::set`? What's an `std::unordered_set`?
+- What's the difference between ordered and unordered containers?
+- What's the difference between `std::list` and `std::forward_list`?
+- How does an `std::array` compare to an `std::vector`? when should you use each?
+- Under what circumstances does the `std::vector` reallocate its buffer?
+- What does `reserve()` do in an `std::vector`?
```

Comment:
I would. What is the difference between `size` and `capacity`

Reply:
Added. Thanks!